src/Init/Data/BitVec/Bitblast.lean

```diff
@@ -1715,6 +1715,120 @@ theorem toInt_sdiv (a b : BitVec w) : (a.sdiv b).toInt = (a.toInt.tdiv b.toInt).
+    obtain ⟨hx₂, hy⟩ := this
+    simp at hy
+    subst hy
+    simp
+    simp at h
+    simp [h]
+  · rintro ⟨hx, hy⟩
+    subst hx
+    subst hy
+    simp
+
+theorem msb_sdiv_eq_decide {x y : BitVec w} :
+    (x.sdiv y).msb = ((0 < w) && 
+      -- x +ve, y +ve: never negative..
+      (!x.msb && y.msb && (-y ≤ x ∧ y ≠ 0#w)) || -- x +ve, y -ve: -ve when result nonzero.
+      (x.msb && !y.msb && (y ≤ -x && y ≠ 0#w)) || -- x -ve, y +ve: -ve when result nonzero.
+      (x.msb && y.msb && (x = intMin w && y = -1#w))) -- x -ve, y -ve:  `intMin / -1 = intMin` -ve.
+     := by
+  by_cases hw : w = 0; subst hw; decide +revert
+  simp only [show 0 < w by omega, decide_true, ne_eq, decide_and, decide_not, Bool.true_and]
+  simp only [sdiv_eq, udiv_eq]
+  rcases hxmsb : x.msb <;> rcases hymsb : y.msb 
+  · -- x:false, y:false
+    simp only [hxmsb, hymsb, msb_udiv_eq_false_of, Bool.not_false, Bool.and_false, Bool.false_and,
+      Bool.and_true, Bool.or_self, Bool.and_self]
+  · -- x:false, y:true
+    simp only [hxmsb, hymsb, msb_neg, msb_udiv_eq_false_of, bne_false, Bool.not_false,
+      Bool.and_self, ne_zero_of_msb_true, decide_false, Bool.and_true, Bool.true_and, Bool.not_true,
+      Bool.false_and, Bool.or_false]
+    simp only[bool_to_prop]
+    have : x / -y ≠ intMin w := by
+        intros h
+        have : (x  / -y).msb = (intMin w).msb := by simp only [h]
+        simp only [msb_intMin, show 0 < w by omega, decide_true] at this
+        rw [msb_udiv] at this
+        simp only [and_eq_true, beq_iff_eq] at this
+        obtain ⟨hcontra, _⟩ := this
+        simp only [hcontra, true_eq_false] at hxmsb
+    simp only [bne_iff_ne, ne_eq, udiv_eq_zero_iff_eq_zero_or_lt, neg_eq_zero_iff,
+      ne_zero_of_msb_true hymsb, _root_.false_or, BitVec.not_lt, this, not_false_eq_true,
+      _root_.and_true]
+  · -- x:true, y:false
+    simp only [hxmsb, hymsb, Bool.not_true, Bool.and_self, Bool.false_and, Bool.not_false,
+      Bool.true_and, Bool.false_or, Bool.and_false, Bool.or_false]
+    by_cases hx₁ : x = 0#w
+    · simp only [hx₁, neg_zero, zero_udiv, msb_zero, le_zero_iff, Bool.and_not_self]
+    · by_cases hy₁ : y = 0#w
+      · simp only [hy₁, udiv_zero, neg_zero, msb_zero, decide_true, Bool.not_true, Bool.and_false]
+      · simp only [hy₁, decide_false, Bool.not_false, Bool.and_true]
+        by_cases hxy₁ : (- x / y) = 0#w
+        · simp only [hxy₁, neg_zero, msb_zero, false_eq_decide_iff, BitVec.not_le]
+          simp only [udiv_eq_zero_iff_eq_zero_or_lt] at hxy₁
+          simp only [hy₁, _root_.false_or] at hxy₁
+          bv_omega
+        · simp only [udiv_eq_zero_iff_eq_zero_or_lt, _root_.not_or, BitVec.not_lt] at hxy₁
+          simp only [hy₁, not_false_eq_true, _root_.true_and] at hxy₁
+          simp only [hxy₁, decide_true]
+          simp only [msb_neg, bne_iff_ne, ne_eq]
+          simp only [bool_to_prop]
+          simp only [bne_iff_ne, ne_eq, udiv_eq_zero_iff_eq_zero_or_lt, hy₁, _root_.false_or,
```

Comment:
can't these be grouped

Reply:
I grouped as much as possible. Sadly still a bit messy.